MatrixSDK/MXSession.m

```diff
@@ -1862,25 +1834,158 @@ - (MXRoom *)directJoinedRoomWithUserId:(NSString*)userId
     return nil;
 }
 
-- (MXHTTPOperation*)uploadDirectRooms:(void (^)(void))success
-                              failure:(void (^)(NSError *error))failure
+
+- (MXHTTPOperation*)setRoom:(NSString*)roomId
+           directWithUserId:(NSString*)userId
+                    success:(void (^)(void))success
+                    failure:(void (^)(NSError *error))failure
 {
+    MXHTTPOperation *operation = [MXHTTPOperation new];
+
     MXWeakify(self);
-    
-    // Push the current direct rooms dictionary to the homeserver.
-    return [self setAccountData:_directRooms forType:kMXAccountDataTypeDirect success:^{
+    [self runOrQueueDirectRoomOperation:^{
         MXStrongifyAndReturnIfNil(self);
-        
-        // Notify a change in direct rooms directory
-        [[NSNotificationCenter defaultCenter] postNotificationName:kMXSessionDirectRoomsDidChangeNotification
-                                                            object:self
-                                                          userInfo:nil];
+
+        NSMutableDictionary<NSString *,NSArray<NSString *> *> *newDirectRooms = [self.directRooms mutableCopy];
+        if (!newDirectRooms)
+        {
+            newDirectRooms = [NSMutableDictionary dictionary];
+        }
+
+        // Remove the current direct user id
+        MXRoom *room = [self roomWithRoomId:roomId];
+        NSString *currentDirectUserId = room.directUserId;
+        if (currentDirectUserId)
+        {
+            NSMutableArray *roomIds = [NSMutableArray arrayWithArray:newDirectRooms[currentDirectUserId]];
+            [roomIds removeObject:roomId];
+
+            if (roomIds.count)
+            {
+                newDirectRooms[currentDirectUserId] = roomIds;
+            }
+            else
+            {
+                [newDirectRooms removeObjectForKey:currentDirectUserId];
+            }
+        }
+
+        // Update with the new one
+        if (userId)
+        {
+            if (![newDirectRooms[userId] containsObject:roomId])
+            {
+                NSMutableArray *roomIds = [NSMutableArray arrayWithArray:newDirectRooms[userId]];
```

Comment:
Here ` newDirectRooms[userId]` may be nil
Don't you think we should use:
`NSMutableArray *roomIds = (newDirectRooms[userId] ? [NSMutableArray arrayWithArray:newDirectRooms[userId]] : [NSMutableArray array]);`